src/Progress/Snapshot.php

```diff
@@ -4,26 +4,95 @@
 namespace BristolSU\Support\Progress;
 
 
+use BristolSU\Support\Activity\Activity;
 use BristolSU\Support\ActivityInstance\ActivityInstance;
 use BristolSU\Support\ActivityInstance\Contracts\ActivityInstanceRepository;
 use BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator;
+use BristolSU\Support\Progress\ProgressUpdateRepository;
 use Carbon\Carbon;
+use Illuminate\Support\Facades\Hash;
 
 class Snapshot
 {
 
+    private ProgressUpdateRepository $ProgressUpdateRepository;
+
+    public function __construct()
+    {
+        $this->ProgressUpdateRepository = new ProgressUpdateRepository();
+    }
+
+    public function createActivityHashArray(Progress $activityProgress)
+    {
+        return [
+            $activityProgress->getActivityInstanceId(),
+            $activityProgress->getPercentage(),
+            json_encode(
+                array_map(fn(ModuleInstanceProgress $moduleInstanceProgress): array => [
+                    'moduleInstanceId' => $moduleInstanceProgress->getModuleInstanceId(),
+                    'mandatory' => $moduleInstanceProgress->isMandatory(),
+                    'complete' => $moduleInstanceProgress->isComplete(),
+                    'percentage' => $moduleInstanceProgress->getPercentage(),
+                    'active' => $moduleInstanceProgress->isActive(),
+                    'visible' => $moduleInstanceProgress->isVisible()
+                ], $activityProgress->getModules())
+            )
+        ];
+    }
+
+    public function ofUpdatesToActivity(Activity $activity, $caller)
+    {
+        $progresses = [];
+        foreach(app(ActivityInstanceRepository::class)->allForActivity($activity->id) as $activityInstance) {
+            array_push($progresses, $this->ofUpdateToActivityInstance($activityInstance, $caller));
+        }
+        return array_filter($progresses);
+    }
+
+    public function ofUpdateToActivityInstance(ActivityInstance $activityInstance, $caller) : ?Progress
+    {
+        // Set Cache key:
+        $itemKey = $caller . '_' . $activityInstance->id;
+
+        // Get the Current Progress:
+        $Progress = $this->ofActivityInstance($activityInstance);
+        $progressArrayHash = $this->createActivityHashArray($Progress);
+
+        $storedProgress = ProgressHashes::find($itemKey);
```

Comment:
As above, this then ties all users of the sdk to use eloquent and a database eloquent supports. Only the repository should have knowledge of how to generate a hash and how to save/retrieve that